packages/core/src/render3/instructions/styling.ts

```diff
@@ -304,322 +159,473 @@ export function ɵɵstyleMap(styles: {[styleName: string]: any} | NO_CHANGE | nu
+  if (stylePropNeedsSanitization(key)) {
+    value = ɵɵsanitizeStyle(value);
+  }
+  arrayMapSet(arrayMap, key, value);
+}
 
-  // we only concat values if there is an initial value, otherwise we return the value as is.
-  // Note that this is to satisfy backwards-compatibility in Angular.
-  if (initialValue.length) {
-    if (isClassBased) {
-      value = concatString(initialValue, forceClassesAsString(bindingValue));
+/**
+ * Update map based styling.
+ *
+ * Map based styling could be anything which contains more than one binding. For example `string`,
+ * `Map`, `Set` or object literal. Dealing with all of these types would complicate the logic so
+ * instead this function expects that the complex input is first converted into normalized
+ * `ArrayMap`. The advantage of normalization is that we get the values sorted, which makes it very
+ * cheap to compute deltas between the previous and current value.
+ *
+ * @param tView Associated `TView.data` contains the linked list of binding priorities.
+ * @param tNode `TNode` where the binding is located.
+ * @param lView `LView` contains the values associated with other styling binding at this `TNode`.
+ * @param renderer Renderer to use if any updates.
+ * @param oldArrayMap Previous value represented as `ArrayMap`
+ * @param newArrayMap Current value represented as `ArrayMap`
+ * @param isClassBased `true` if `class` (`false` if `style`)
+ * @param bindingIndex Binding index of the binding.
+ */
+function updateStylingMap(
+    tView: TView, tNode: TNode, lView: LView, renderer: Renderer3, oldArrayMap: ArrayMap<any>,
+    newArrayMap: ArrayMap<any>, isClassBased: boolean, bindingIndex: number) {
+  if (oldArrayMap as ArrayMap<any>| NO_CHANGE === NO_CHANGE) {
+    // ON first execution the oldArrayMap is NO_CHANGE => treat is as empty ArrayMap.
+    oldArrayMap = EMPTY_ARRAY as any;
+  }
+  let oldIndex = 0;
+  let newIndex = 0;
+  let oldKey: string|null = 0 < oldArrayMap.length ? oldArrayMap[0] : null;
+  let newKey: string|null = 0 < newArrayMap.length ? newArrayMap[0] : null;
+  while (oldKey !== null || newKey !== null) {
+    ngDevMode && assertLessThan(oldIndex, 999, 'Are we stuck in infinite loop?');
+    ngDevMode && assertLessThan(newIndex, 999, 'Are we stuck in infinite loop?');
+    const oldValue = oldIndex < oldArrayMap.length ? oldArrayMap[oldIndex + 1] : undefined;
+    const newValue = newIndex < newArrayMap.length ? newArrayMap[newIndex + 1] : undefined;
+    let setKey: string|null = null;
+    let setValue: any = undefined;
+    if (oldKey === newKey) {
+      // UPDATE: Keys are equal => new value is overwriting old value.
+      oldIndex += 2;
+      newIndex += 2;
+      if (oldValue !== newValue) {
+        setKey = newKey;
+        setValue = newValue;
+      }
+    } else if (newKey === null || oldKey !== null && oldKey < newKey !) {
+      // DELETE: oldKey key is missing or we did not find the oldKey in the newValue.
```

Comment:
It took me a sec to realize that you were comparing first letters, so comment might save others time

```suggestion
// DELETE: oldKey key is missing or we did not find the oldKey in the newValue 
// (because the arrayMap is sorted and `newKey` has a later first letter).
```